fix: FIT-1073: Deleted all tasks but one in Done, project still in Annotating

matt-bernstein opened the pull request:
FSM bugfix
<!--

This description MUST be filled out for a PR to receive a review. Its primary purposes are:

 - to enable your reviewer to review your code easily, and
 - to convince your reviewer that your code works as intended.

Some pointers to think about when filling out your PR description:
 - Reason for change: Description of problem and solution
 - Screenshots: All visible changes should include screenshots.
 - Rollout strategy: How will this code be rolled out? Feature flags / env var / other
 - Testing: Description of how this is being verified
 - Risks: Are there any known risks associated with this change, eg to security or performance?
 - Reviewer notes: Any info to help reviewers approve the PR
 - General notes: Any info to help onlookers understand the code, or callouts to significant portions.

You may use AI tools such as Copilot Actions to assist with writing your PR description (see https://docs.github.com/en/copilot/using-github-copilot/using-github-copilot-for-pull-requests/creating-a-pull-request-summary-with-github-copilot); however, an AI summary isn't enough by itself. You'll need to provide your reviewer with strong evidence that your code works as intended, which requires actually running the code and showing that it works.

-->

matt-bernstein committed:
fix: FIT-1073: Deleted all tasks but one in Done, project still in Annotating

matt-bernstein committed:
Merge remote-tracking branch 'origin/develop' into fb-FIT-1073

bmartel commented on label_studio/tasks/models.py:
```diff
@@ -1312,6 +1312,11 @@ def update_all_task_states_after_deleting_task(sender, instance, **kwargs):
+        from core.utils.common import load_func
+
+        update_func = load_func(settings.FSM_UPDATE_PROJECT_STATE_AFTER_TASK_CHANGE)
+        update_func(instance.project, user=getattr(instance, 'updated_by', None))
```
We already had an update running here, I guess the same question as above?

bmartel commented on label_studio/data_manager/actions/basic.py:
```diff
@@ -58,6 +62,9 @@ def delete_tasks(project, queryset, **kwargs):
     )
+    # FSM: Recalculate project state after deletion/unlink using swappable setting
+    update_func = load_func(settings.FSM_UPDATE_PROJECT_STATE_AFTER_TASK_CHANGE)
+    update_func(project, user=user)
```
We already had a project update running here, is there something wrong with the `update_tasks_states` function? Is that an FSM-based function, or what is that doing?

matt-bernstein replied:
It predates FSM, it updates task counters in the project. It also runs async in Redis, so I need to do some thinking about whether this new update is best done sync or async.

matt-bernstein committed:
fix test

matt-bernstein committed:
Merge remote-tracking branch 'origin/develop' into fb-FIT-1073

matt-bernstein committed:
move recalc to async job

matt-bernstein committed:
Merge remote-tracking branch 'origin/develop' into fb-FIT-1073

matt-bernstein commented:
/git merge develop
> [Workflow run](https://github.com/HumanSignal/label-studio/actions/runs/19899045852)
> Successfully merged:  2 files changed, 32 insertions(+), 5 deletions(-)

matt-bernstein committed:
Merge branch 'develop' into 'fb-FIT-1073'

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/19899045852

matt-bernstein commented:
/git merge develop
> [Workflow run](https://github.com/HumanSignal/label-studio/actions/runs/19900610982)
> Successfully merged:  6 files changed, 106 insertions(+), 33 deletions(-)

matt-bernstein committed:
Merge branch 'develop' into 'fb-FIT-1073'

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/19900610982

matt-bernstein commented:
/git merge develop
> [Workflow run](https://github.com/HumanSignal/label-studio/actions/runs/19906717575)
> Successfully merged:  create mode 100644 docs/themes/v2/source/images/releases/flight_app_builder.gif

matt-bernstein committed:
Merge branch 'develop' into 'fb-FIT-1073'

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/19906717575

bmartel commented on label_studio/projects/models.py:
```diff
@@ -529,6 +529,11 @@ def _update_tasks_states(
+        if tasks_number_changed:
+            # FSM: Recalculate project state after task deletion or import
+            update_func = load_func(settings.FSM_UPDATE_PROJECT_STATE_AFTER_TASK_CHANGE)
+            update_func(self, user=None)
```
Can we move the load_func to a module level definition with a name like `update_project_state_after_task_change = load_func(settings.FSM_UPDATE_PROJECT_STATE_AFTER_TASK_CHANGE)`

matt-bernstein committed:
make FSM state update a utility fn, plumb user through

matt-bernstein committed:
pass user to update_tasks_states and update_tasks_counters_and_task_states everywhere

matt-bernstein commented:
/fmt
> [Workflow run](https://github.com/HumanSignal/label-studio/actions/runs/19908358800)

matt-bernstein committed:
Apply pre-commit linters

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/19908358800

matt-bernstein committed:
bugfix

matt-bernstein commented:
/fmt
> [Workflow run](https://github.com/HumanSignal/label-studio/actions/runs/19908527354)

matt-bernstein commented:
Realized this is more complex than I thought while using the async job - have to plumb `user` through all usages of it to set the FSM transition history correctly.

matt-bernstein committed:
Apply pre-commit linters

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/19908527354

bmartel commented on label_studio/fsm/project_transitions.py:
```diff
@@ -154,3 +156,8 @@ def update_project_state_after_task_change(project, user=None):
             StateManager.execute_transition(entity=project, transition_name='project_completed', user=user)
+
+
+def update_project_state_after_task_change(project, user=None):
+    update_func = load_func(settings.FSM_UPDATE_PROJECT_STATE_AFTER_TASK_CHANGE)
+    return update_func(project, user)
```
```suggestion
def update_project_state_after_task_change(project):
    user = CurrentContext.get_user() # <-- The concept of current user is already plumbed for you, even across jobs!
    update_func = load_func(settings.FSM_UPDATE_PROJECT_STATE_AFTER_TASK_CHANGE)
    return update_func(project, user)
```

matt-bernstein replied:
Right now I'm passing CurrentContext in the async job, not in here. If it gets passed here, it may simplify the overall FSM code but it'll explode this diff with dozens of usages of `update_project_state_after_task_change
(project, user=user)` -> `update_project_state_after_task_change
(project)`. Possibly worth it? I'm also wondering about the perf impact of re-calling CurrentContext() in a place where the caller of this function also may have done so, including inside some hot loops.

matt-bernstein committed:
Merge remote-tracking branch 'origin/develop' into fb-FIT-1073

matt-bernstein commented:
/fmt
> [Workflow run](https://github.com/HumanSignal/label-studio/actions/runs/19932483747)

matt-bernstein committed:
Revert "pass user to update_tasks_states and update_tasks_counters_and_task_states everywhere"

This reverts commit 8f92a5e954e739eb7b5a6d33e911d2eb19d8f8b8.

matt-bernstein committed:
use CurrentContext for user in _update_tasks_states

matt-bernstein commented:
/git merge develop
> [Workflow run](https://github.com/HumanSignal/label-studio/actions/runs/19938300811)
> Successfully merged:  2 files changed, 55 insertions(+), 51 deletions(-)

matt-bernstein committed:
Merge branch 'develop' into 'fb-FIT-1073'

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/19938300811

matt-bernstein commented:
/git merge develop
> [Workflow run](https://github.com/HumanSignal/label-studio/actions/runs/19940461456)
> Successfully merged:  2 files changed, 30 insertions(+), 3 deletions(-)

matt-bernstein committed:
Merge branch 'develop' into 'fb-FIT-1073'

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/19940461456

matt-bernstein commented:
/git merge develop
> [Workflow run](https://github.com/HumanSignal/label-studio/actions/runs/19941523519)
> Successfully merged:  1 file changed, 3 insertions(+), 1 deletion(-)

matt-bernstein committed:
Merge branch 'develop' into 'fb-FIT-1073'

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/19941523519

matt-bernstein commented:
/git merge develop
> [Workflow run](https://github.com/HumanSignal/label-studio/actions/runs/19944949964)
> Successfully merged:  2 files changed, 4 insertions(+), 1 deletion(-)

matt-bernstein committed:
Merge branch 'develop' into 'fb-FIT-1073'

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/19944949964

matt-bernstein commented:
/git merge develop
> [Workflow run](https://github.com/HumanSignal/label-studio/actions/runs/19966654403)
> Successfully merged:  4 files changed, 61 insertions(+), 5 deletions(-)

matt-bernstein committed:
Merge branch 'develop' into 'fb-FIT-1073'

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/19966654403

matt-bernstein commented:
/git merge develop
> [Workflow run](https://github.com/HumanSignal/label-studio/actions/runs/19975977679)
> Successfully merged:  3 files changed, 12 insertions(+), 12 deletions(-)

matt-bernstein committed:
Merge branch 'develop' into 'fb-FIT-1073'

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/19975977679

yyassi-heartex committed:
making sure we resepect context fsm state

matt-bernstein commented:
/git merge develop
> [Workflow run](https://github.com/HumanSignal/label-studio/actions/runs/20030395775)
> Successfully merged:  create mode 100644 web/libs/ui/src/assets/icons/messages-check.svg

matt-bernstein committed:
Merge branch 'develop' into 'fb-FIT-1073'

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/20030395775

matt-bernstein commented:
/git merge develop
> [Workflow run](https://github.com/HumanSignal/label-studio/actions/runs/20035791286)
> Successfully merged:  2 files changed, 34 insertions(+), 12 deletions(-)

matt-bernstein committed:
Merge branch 'develop' into 'fb-FIT-1073'

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/20035791286

yyassi-heartex committed:
Merge remote-tracking branch 'origin/develop' into fb-FIT-1073
